packages/kbn-alerts-grouping/src/alerts_grouping/components/alerts_grouping_level.tsx

```diff
@@ -0,0 +1,206 @@
+  ) => ReactElement;
+  groupingLevel?: number;
+  onGroupClose: () => void;
+  pageIndex: number;
+  pageSize: number;
+  parentGroupingFilter?: string;
+  selectedGroup: string;
+  setPageIndex: (newIndex: number) => void;
+  setPageSize: (newSize: number) => void;
+}
+
+const DEFAULT_FILTERS: Filter[] = [];
+
+/**
+ * Renders an alerts grouping level
+ */
+export const AlertsGroupingLevel = memo(
+  ({
+    featureIds,
+    defaultFilters = DEFAULT_FILTERS,
+    from,
+    getGrouping,
+    globalFilters,
+    globalQuery,
+    groupingLevel,
+    loading = false,
+    onGroupClose,
+    pageIndex,
+    pageSize,
+    parentGroupingFilter,
+    children,
+    selectedGroup,
+    setPageIndex,
+    setPageSize,
+    to,
+    takeActionItems,
+    getAggregationsByGroupingField,
+    services: { http, notifications },
+  }: AlertsGroupingLevelProps) => {
+    const additionalFilters = useMemo(() => {
+      try {
+        return [
+          buildEsQuery(undefined, globalQuery != null ? [globalQuery] : [], [
+            ...(globalFilters?.filter((f) => f.meta.disabled === false) ?? []),
+            ...(defaultFilters ?? []),
+            ...(parentGroupingFilter ? JSON.parse(parentGroupingFilter) : []),
+          ]),
+        ];
+      } catch (e) {
+        return [];
+      }
+    }, [defaultFilters, globalFilters, globalQuery, parentGroupingFilter]);
+
+    // Create a unique, but stable (across re-renders) value
+    const uniqueValue = useMemo(() => `SuperUniqueValue-${uuidv4()}`, []);
+
+    const queryGroups = useMemo(() => {
+      return getAlertsGroupingQuery({
+        additionalFilters,
+        selectedGroup,
+        uniqueValue,
+        from,
+        to,
+        pageSize,
+        pageIndex,
+        getAggregationsByGroupingField,
+      });
+    }, [
+      additionalFilters,
+      from,
+      getAggregationsByGroupingField,
+      pageIndex,
+      pageSize,
+      selectedGroup,
+      to,
+      uniqueValue,
+    ]);
+
+    const [finalQuery, setFinalQuery] = useState(queryGroups);
+
+    const { data: alertGroupsData, isLoading: isLoadingGroups } = useFindAlertsQuery<
+      GroupingAggregation<AlertsGroupingAggregation>
+    >({
+      http,
+      toasts: notifications.toasts,
+      enabled: finalQuery && !isNoneGroup([selectedGroup]),
+      params: {
+        feature_ids: featureIds,
+        ...finalQuery,
+      },
+    });
+
+    const queriedGroup = useRef<string | null>(null);
+
+    const aggs = useMemo(
+      // queriedGroup because `selectedGroup` updates before the query response
+      () =>
+        parseGroupingQuery(
+          // fallback to selectedGroup if queriedGroup.current is null, this happens in tests
+          queriedGroup.current === null ? selectedGroup : queriedGroup.current,
+          uniqueValue,
+          alertGroupsData?.aggregations
+        ),
+      [alertGroupsData?.aggregations, selectedGroup, uniqueValue]
+    );
+
+    useEffect(() => {
```

Comment:
I think we can avoid this `useEffect` by removing the `useState(queryGroups)` and derive the state like

```
const finalQuery = !isNoneGroup([selectedGroup]) ? {...queryGroups, current: queryGroups?.runtime_mappings?.groupByField?.script?.params?.selectedGroup ?? ''} : queryGroups

const { data: alertGroupsData, isLoading: isLoadingGroups } = useFindAlertsQuery<
      GroupingAggregation<AlertsGroupingAggregation>
    >({
      http,
      toasts: notifications.toasts,
      enabled: finalQuery && !isNoneGroup([selectedGroup]),
      params: {
        feature_ids: featureIds,
        ...finalQuery,
      },
    });
```

Wdyt?

Reply:
Absolutely agree, done in https://github.com/elastic/kibana/pull/184635/commits/c0cb7a1d42e09b0c66e66d8c841833558d40faa3